### PR TITLE
manifests: accept optional parameter functions

### PR DESCRIPTION
**Pull request by Tal-or**
We want to have the option to set/create the namespace and/or set the objects' names.
A good way to allow this flexibility is by using optional parameter functions.

When the client controls the name,
it can store the name and delete the objects without having the need to hold the objects themselves.
    
In our case for example we want to be able to
the node-plugin (delete the objects) without deleting the profile.
    
This can be done easily by specifying the name
of the objects we want to delete.

### CODE REVIEW

File: pkg/manifests/manifests.go [hunk 1]
```diff
@@ -74,36 +74,56 @@ func Get(ns string, sharedCPUs string) (*Manifests, error) {
...
+		fmt.Sprintf("system:serviceaccount:%s:%s", mf.SA.Namespace, mf.SA.Name),
+	}
+	if err := mf.SetSharedCPUs(sharedCPUs); err != nil {
+		return nil, err
+	}
+	return &mf, nil
+}
+
+// WithNewNamespace creates new namespace and updates the objects
+func WithNewNamespace(ns string) func(mf *Manifests) {
```
Comment (ffromani) [inline pkg/manifests/manifests.go:95]: this seems like mixing layers (setting properties AND creating objects) and, albeit likely handy, I'd avod it

File: pkg/manifests/manifests.go [hunk 2]
```diff
@@ -74,36 +74,56 @@ func Get(ns string, sharedCPUs string) (*Manifests, error) {
 		}
 	}
 
-	if ns != "" {
+	// by default, set the namespace to default namespace.
+	// the default namespace can be changed via Optional Parameter Function
+	WithNamespace("")(&mf)
```
Comment (ffromani) [inline pkg/manifests/manifests.go:79]: this is awkward to read though. I'd extract a common helper or even duplicate the namespace setting.

File: pkg/manifests/manifests.go [hunk 3]
```diff
@@ -74,36 +74,56 @@ func Get(ns string, sharedCPUs string) (*Manifests, error) {
...
-		&mf.NS,
+	obj := []client.Object{
 		&mf.DS,
 		&mf.Role,
 		&mf.RB,
 		&mf.SA,
 		&mf.SCC,
 	}
+	if mf.NS.Name != "" {
+		obj = append(obj, &mf.NS)
```
Comment (ffromani) [inline pkg/manifests/manifests.go:135]: append or prepend? is the order of objects meaningful, or should it be?

File: pkg/manifests/manifests.go [hunk 4]
```diff
@@ -109,6 +109,17 @@ func WithNamespace(ns string) func(mf *Manifests) {
 		mf.Role.Namespace = ns
 		mf.RB.Namespace = ns
 		mf.SA.Namespace = ns
+		// SCC is cluster scoped
+	}
+}
+
+func WithName(name string) func(mf *Manifests) {
+	return func(mf *Manifests) {
```
Comment (ffromani) [inline pkg/manifests/manifests.go:117]: but do we _want_ to have the same name for all the objects? what's the rationale?

**Review by ffromani — COMMENTED** [review]
"make extensibile because we may need it later" is [not a very compelling argument](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it)

The main practical problem is we design an API for usecases we are not clear yet (otherwise is not "we may need it" but rather "we do need it"), and this has a cost later to undo the abstraction if we guessed wrong.

**Comment by Tal-or** [comment]
> "make extensibile because we may need it later" is [not a very compelling argument](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it)
> 
> The main practical problem is we design an API for usecases we are not clear yet (otherwise is not "we may need it" but rather "we do need it"), and this has a cost later to undo the abstraction if we guessed wrong.

We do need it. this was a preparation for adding the `WithName()` function. the `WithName()` is useful for deployment, because the client controls the name, hence it's easier to delete.

File: pkg/manifests/manifests.go (hunk 1 above)
Comment (Tal-or) [reply pkg/manifests/manifests.go:95]: How do you suggest handle the namespace creation?

**Comment by ffromani** [comment]
> > "make extensibile because we may need it later" is [not a very compelling argument](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it)
> > The main practical problem is we design an API for usecases we are not clear yet (otherwise is not "we may need it" but rather "we do need it"), and this has a cost later to undo the abstraction if we guessed wrong.
> 
> We do need it. this was a preparation for adding the `WithName()` function. the `WithName()` is useful for deployment, because the client controls the name, hence it's easier to delete. And we are not always need the name so it's optional

great, so it's just a matter to rephrase the [PR cover letter](https://github.com/openshift-kni/mixed-cpu-node-plugin/pull/10#issue-1769369313) (and possibly some commit messages)

File: pkg/manifests/manifests.go (hunk 4 above)
Comment (Tal-or) [reply pkg/manifests/manifests.go:117]: When the client controls the name, it can store the name and delete the objects without having the need to hold the objects themselves.

**Comment by Tal-or** [comment]
> > > "make extensibile because we may need it later" is [not a very compelling argument](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it)
> > > The main practical problem is we design an API for usecases we are not clear yet (otherwise is not "we may need it" but rather "we do need it"), and this has a cost later to undo the abstraction if we guessed wrong.
> > 
> > 
> > We do need it. this was a preparation for adding the `WithName()` function. the `WithName()` is useful for deployment, because the client controls the name, hence it's easier to delete. And we are not always need the name so it's optional
> 
> great, so it's just a matter to rephrase the [PR cover letter](https://github.com/openshift-kni/mixed-cpu-node-plugin/pull/10#issue-1769369313) (and possibly some commit messages)

Yea I would do that. Thanks!

File: pkg/manifests/manifests.go (hunk 4 above)
Comment (ffromani) [reply pkg/manifests/manifests.go:117]: I think this is ok and has no ill effect but now I'm thinking if this is the best way to achieve this goal. IIRC using the `ControllerReference` is the best/preferred approach.
TL;DR: I'm fine with using the common name, but let's make sure address the actual need (make it easy to delete objects in the right order without leftovers)

Comment (Tal-or) [reply pkg/manifests/manifests.go:117]: In our case for example we want to be able to disable the feature (delete the objects) without deleting the profile

Comment (ffromani) [reply pkg/manifests/manifests.go:117]: This is a very good point and deserves a clear comment in the codebase. So we need a partial reimplementation of `ControllerReference`. I'm wondering if we can add an annotation, or a `ObjectReference` to get the same functionality. This can totally deferred, but I'd rather look into this sooner or later to look for the best practices according to k8s standards, and I'd check if there are precedents or similar enough cases which can help.
For now I recommend to move on with the common name per your approach and add a note to check for the future.

Comment (Tal-or) [reply pkg/manifests/manifests.go:119]: This sounds more and more like it deserves a fully functional controller of it's own :)

**Comment by Tal-or** [comment]
depends on #11

**Comment by Tal-or** [comment]
@ffromani Any other comments from your side or shall we continue with merging?

**Review by ffromani — COMMENTED** [review]
/lgtm

looks good enough for merging (and iterate later if needed) thanks
